### PR TITLE
WIP: mutating webhook for injecting java agent to pods

### PR DESCRIPTION
**Pull request by ravisingal**

### CODE REVIEW

File: webhook/Makefile [hunk 1]
```diff
@@ -0,0 +1,18 @@
+NAME = javaagent-webhook
+IMAGE_PREFIX = hypertrace
```
Comment (pavolloffay) [inline webhook/Makefile:2]: nit: `DOCKER_NAMESPACE`

also make all the env vars configurable - `?= hypertrace`

File: webhook/Makefile [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+	go build -v -o $(NAME) ./cmd/
+
+deps:
+	go get -v ./...
+
+test: deps
+	go test -v ./... -cover
+
+docker:
+	docker build --no-cache -t $(IMAGE_PREFIX)/$(NAME):latest .
```
Comment (pavolloffay) [inline webhook/Makefile:16]: make the tag also configurable

File: webhook/go.mod [hunk 3]
```diff
@@ -0,0 +1,9 @@
+module github.com/hypertrace/javaagent-webhook
+
+go 1.12
```
Comment (pavolloffay) [inline webhook/go.mod:3]: could we start with go 15?

File: webhook/Makefile [hunk 4]
```diff
@@ -0,0 +1,18 @@
+NAME = javaagent-webhook
+IMAGE_PREFIX = hypertrace
+
+export GO111MODULE=on
```
Comment (pavolloffay) [inline webhook/Makefile:4]: nit: not needed as we should use go 15 from start

File: webhook/.go-version [hunk 5]
```diff
@@ -0,0 +1 @@
+1.15.6
```
Comment (pavolloffay) [inline webhook/.go-version:1]: does `go` use this file?

File: webhook/cmd/webhook.go [hunk 6]
```diff
@@ -0,0 +1,238 @@
...
+	var value interface{}
+
+	if len(pod.Spec.Volumes) != 0 {
+		path = path + "/-"
+		value = volume
+	} else {
+		value = []corev1.Volume{volume}
+	}
+
+	patch = append(patch, patchOperation{
```
Comment (pavolloffay) [inline webhook/cmd/webhook.go:56]: nit: return directly

File: webhook/cmd/webhook.go [hunk 7]
```diff
@@ -0,0 +1,238 @@
...
+// add volumeMount and JAVA_TOOL_OPTIONS environment variable
+func updateContainers(pod *corev1.Pod) (patch []patchOperation) {
+
+	modifiedContainers := []corev1.Container{}
+	serviceName := getServiceName(pod)
+
+	for _, container := range pod.Spec.Containers {
+		ignore := false
+		for _, envVar := range container.Env {
+			if envVar.Name == "HYPERTRACE_IGNORE_JAVAAGENT" {
```
Comment (pavolloffay) [inline webhook/cmd/webhook.go:86]: we should probably use a label for this.

File: webhook/cmd/webhook_test.go [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+	"encoding/json"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+
+	v1beta1 "k8s.io/api/admission/v1beta1"
+)
+
+func TestMutatesValidRequest(t *testing.T) {
+	rawJSON := `{
```
Comment (pavolloffay) [inline webhook/cmd/webhook_test.go:13]: could we test only parts that the code changes?

File: webhook/cmd/webhook.go [hunk 9]
```diff
@@ -0,0 +1,238 @@
...
+			for _, envVar := range container.Env {
+				if envVar.Name == "JAVA_TOOL_OPTIONS" {
+					envVar.Value = envVar.Value + " -javaagent:/mnt/hypertrace/hypertrace-agent-all.jar"
+					found = true
+				}
+				modifiedEnvVars = append(modifiedEnvVars, envVar)
+			}
+			if !found {
+				modifiedEnvVars = append(modifiedEnvVars, corev1.EnvVar{
+					Name:  "JAVA_TOOL_OPTIONS",
```
Comment (pavolloffay) [inline webhook/cmd/webhook.go:112]: nit: extract these to constants

File: webhook/cmd/webhook.go [hunk 10]
```diff
@@ -0,0 +1,238 @@
...
+					Value: "-javaagent:/mnt/hypertrace/hypertrace-agent-all.jar",
+				})
+			}
+
+			if serviceName != "" {
+				modifiedEnvVars = append(modifiedEnvVars, corev1.EnvVar{
+					Name:  "HT_SERVICE_NAME",
```
Comment (pavolloffay) [inline webhook/cmd/webhook.go:119]: nit: extract to constants also all env names

File: deploy/webhook.yaml [hunk 11]
```diff
@@ -0,0 +1,111 @@
+---
```
Comment (pavolloffay) [inline deploy/webhook.yaml:1]: I would move this file into `./webhook` dir since it's related to webhook deployment

**Review by pavolloffay — COMMENTED** [review]
it looks good ❤️ , mostly nit comments.

* we should add a CI job to run at least the tests
* we should add readme into webhook dir to explain the approach and how to build it.

File: webhook/cmd/webhook.go (hunk 7 above)
Comment (ravisingal) [reply webhook/cmd/webhook.go:86]: we needed it at container level. labels are pod specific. This is added not to inject java-agent to one of the container in multi-containers pod.

Comment (pavolloffay) [reply webhook/cmd/webhook.go:86]: from the UX perspective we need these features: inject agent if a lebel is present e.g. on namespace/deployment/pod. Exclude specific deployments/pods if injection is enabled on a namespace. 

We don't have to implement these right away, we just should know that it is feasible.

Comment (pavolloffay) [reply webhook/cmd/webhook.go:86]: We will also need to provide agent configuration in config map - see `agent-attacher-config-map` in traceableai nameslace. This CM embeds the hypertrace agent config that the agent uses.

Comment (pavolloffay) [reply webhook/cmd/webhook.go:86]:
```
k get configmap agent-attacher-config-map -n traceableai -o json | jq '.data."agent-attacher-config.yaml"' -r                                                                                                       1 ↵ ploffay@Pavols-MBP
agent_attacher:
  logging:
    level: info
    encoding: "json"
    outputPaths:
      - stdout
    errorOutputPaths:
      - stderr
    encoderConfig:
      timeKey: "time"
      timeEncoder: "iso8601"
      levelKey: "level"
      levelEncoder: "uppercase"
      messageKey: "message"
  reporting:
    address: api-staging.traceable.ai
  kubernetes:
    auto_instrument_services: false
    namespaces:
      - label_selector:
          traceableai-instrumentation=enabled
        attach_disabled: false
        instrumentation_agent:
          use_hypertrace_agent: true
          hypertrace_agent_config:
            reporting:
              endpoint: "http://collector.traceableai:9411/api/v2/spans"
            opa:
              endpoint: "http://opa.traceableai:8181/"
            dataCapture:
              httpHeaders:
                request: true
                response: true
              httpBody:
                request: true
                response: true
              rpcMetadata:
                request: true
                response: true
              rpcBody:
                request: true
                response: true
          logging:
            level: info
            outputPaths:
              - /tmp
            errorOutputPaths:
              - /tmp
          reporting:
            address: api-staging.traceable.ai
            traces_address: "http://collector.traceableai:14268/api/traces"
          java_agent:
            jattach: true
          disable_capture_of_attributes:
            - http.request.body
            - http.request.header
            - http.response.body
            - http.response.header

    pods:
      - label_selector:
          traceableai-instrumentation=enabled
        attach_disabled: false
        instrumentation_agent:
          use_hypertrace_agent: true
          hypertrace_agent_config:
            reporting:
              endpoint: "http://collector.traceableai:9411/api/v2/spans"
            opa:
              endpoint: "http://opa.traceableai:8181/"
            dataCapture:
              httpHeaders:
                request: true
                response: true
              httpBody:
                request: true
                response: true
              rpcMetadata:
                request: true
                response: true
              rpcBody:
                request: true
                response: true
          logging:
            level: info
            outputPaths:
              - /tmp
            errorOutputPaths:
              - /tmp
          reporting:
            address: api-staging.traceable.ai
            traces_address: "http://collector.traceableai:14268/api/traces"
          java_agent:
            jattach: true
          disable_capture_of_attributes:
            - http.request.body
            - http.request.header
            - http.response.body
            - http.response.header
      - label_selector:
          traceableai-instrumentation=disabled
        attach_disabled: true

```

File: webhook/Makefile (hunk 1 above)
Comment (ravisingal) [reply webhook/Makefile]: fixed

File: webhook/Makefile (hunk 2 above)
Comment (ravisingal) [reply webhook/Makefile]: fixed

File: webhook/go.mod (hunk 3 above)
Comment (ravisingal) [reply webhook/go.mod]: fixed

File: webhook/Makefile (hunk 4 above)
Comment (ravisingal) [reply webhook/Makefile]: fixed

File: webhook/.go-version (hunk 5 above)
Comment (ravisingal) [reply webhook/.go-version]: removed file.

File: webhook/cmd/webhook.go (hunk 6 above)
Comment (ravisingal) [reply webhook/cmd/webhook.go]: fixed

File: webhook/cmd/webhook.go (hunk 7 above)
Comment (ravisingal) [reply webhook/cmd/webhook.go:84]: configmaps are namespace scoped and to mount it on a pod, we have to make sure that the configmap exists in the namespace of the pod. Instead, we can mount this configmap on the webhook pod and copy the YAML file to the same shared volume where the java agent jar is copied.

File: deploy/webhook.yaml [hunk 12]
```diff
@@ -0,0 +1,111 @@
...
+spec:
+  replicas: 1
+  selector:
+    matchLabels:
+      app: hypertrace-javaagent-webhook
+  template:
+    metadata:
+      name: hypertrace-javaagent-webhook
+      annotations:
+        sidecar.istio.io/inject: "false"
```
Comment (mohit-a21) [inline deploy/webhook.yaml:55]: I am not sure if this is needed.

File: deploy/webhook.yaml [hunk 13]
```diff
@@ -0,0 +1,111 @@
...
+        - name: ssl-cert
+          secret:
+            secretName: hypertrace-javaagent-webhook-certs
+---
+apiVersion: admissionregistration.k8s.io/v1beta1
+kind: MutatingWebhookConfiguration
+metadata:
+  name: hypertrace-javaagent-webhook
+  annotations:
+    cert-manager.io/inject-ca-from: hypertrace/hypertrace-javaagent-webhook-certs
```
Comment (ravisingal) [inline deploy/webhook.yaml:89]: This requires cert-manager helm chart - https://github.com/jetstack/cert-manager/tree/master/deploy/charts/cert-manager:

File: webhook/cmd/webhook.go (hunk 7 above)
Comment (ravisingal) [reply webhook/cmd/webhook.go:84]: @pavolloffay @davexroth @mohit-a21 hypertrace javaagent needs the whole configmap or just the section under `hypertrace_agent_config`?

Comment (pavolloffay) [reply webhook/cmd/webhook.go:84]: just the section under `hypertrace_agent_config`. The config proto is here https://github.com/hypertrace/agent-config.

**Comment by ravisingal** [comment]
not required anymore.

**Pull request closed without merging**